engine/src/main/java/io/camunda/zeebe/engine/state/compensation/DbCompensationSubscriptionState.java

```diff
@@ -106,7 +106,8 @@ public Set<CompensationSubscription> findSubscriptionsByThrowEventInstanceKey(
     compensationSubscriptionColumnFamily.whileEqualPrefix(
         new DbCompositeKey<>(tenantIdKey, processInstanceKey),
         ((key, value) -> {
-          if (value.getRecord().getThrowEventInstanceKey() == throwEventInstanceKey) {
+          if (value.getRecord().getThrowEventInstanceKey() == throwEventInstanceKey
+              && !value.getRecord().getCompensationHandlerId().isEmpty()) {
```

Comment:
> Why is this bad? Since the subscription is not deleted, it pollutes the state (i.e. memory leak). In a long-running process, the subscriptions could lead to a big state.

We delete the subprocess subscription from the state on termination/completion of the process. So the state is not polluted by the subscriptions.

> 
> Better: delete the subscriptions of the subprocesses from the state. When all subscriptions of a subprocess are completed, we complete the subscription of the subprocess as well.

Is this still valid ? I think yes, because we should complete the subprocess subscriptions as well

Reply:
> We delete the subprocess subscription from the state on termination/completion of the process.

True. But if the process instance is active for a long time (i.e. a long-running process) then the subscriptions are not removed from the state and the state grows.

The modeling is up to the users. We know cases where process instances are active for days, weeks, months, and more. :sweat_smile:

Reply:
I see, thanks for raising this